lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -405,6 +412,313 @@ bool ADContext::supportsFloatingPointDifferentiation(Type type) const {
+// Activity Analysis
+//===----------------------------------------------------------------------===//
+
+/// In many real situations, the end-users of AD need only the derivatives of
+/// some selected outputs of `P` with respect to some selected inputs of `P`.
+/// Whatever the differentiation mode (tangent, reverse,...), these restrictions
+/// allow the AD tool to produce a much more efficient differentiated program.
+/// Essentially, fixing some inputs and neglecting some outputs allows AD to
+/// just forget about several intermediate differentiated variables.
+///
+/// Activity analysis is the specific analysis that detects these situations,
+/// therefore allowing for a better differentiated code. Activity analysis is
+/// present in all transformation-based AD tools.
+///
+/// To begin with, the end-user specifies that only some output variables (the
+/// “dependent”) must be differentiated with respect to only some input
+/// variables (the “independent”). We say that variable `y` depends on `x` when
+/// the derivative of `y` with respect to `x` is not trivially null. We say that
+/// a variable is “varied” if it depends on at least one independent. Conversely
+/// we say that a variable is “useful” if at least one dependent depends on it.
+/// Finally, we say that a variable is “active” if it is at the same time varied
+/// and useful. In the special case of the tangent mode, it is easy to check
+/// that when variable `v` is not varied at some place in the program, then its
+/// derivative `v̇` at this place is certainly null. Conversely when variable `v`
+/// is not useful, then whatever the value of `v̇`, this value does not matter
+/// for the final result. Symmetric reasoning applies for the reverse mode of
+/// AD: observing that differentiated variables go upstream, we see that a
+/// useless variable has a null derivative, in other words the partial
+/// derivative of the output with respect to this variable is null. Conversely
+/// when variable `v` is not varied, then whatever the value of `v`, this value
+/// does not matter for the final result.
+///
+/// Reference:
+/// Laurent Hascoët. Automatic Differentiation by Program Transformation. 2017.
+
+namespace {
+
+class DifferentiableActivityInfo;
+
+class DifferentiableActivityAnalysis
+  : public FunctionAnalysisBase<DifferentiableActivityInfo> {
+private:
+  DominanceAnalysis *dominanceAnalysis = nullptr;
+
+public:
+  explicit DifferentiableActivityAnalysis()
+    : FunctionAnalysisBase(AnalysisKind::DifferentiableActivity) {}
+
+  static bool classof(const SILAnalysis *s) {
+    return s->getKind() == AnalysisKind::DifferentiableActivity;
+  }
+
+  virtual bool shouldInvalidate(SILAnalysis::InvalidationKind k) override {
+    return k & InvalidationKind::Everything;
+  }
+
+  virtual
+  DifferentiableActivityInfo *newFunctionAnalysis(SILFunction *f) override;
+
+  virtual void initialize(SILPassManager *pm) override;
+};
+
+class DifferentiableActivityInfo {
```

Comment:
Please add a doc comment even if short.  Please shrink wrap the anonymous namespace around the classes, in accordance to llvm.org/docs/CodingStandards.html